src/frequenz/sdk/timeseries/_formula_engine/_formula_printer.py

```diff
@@ -0,0 +1,220 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+from __future__ import annotations
+import enum
+
+from ._formula_steps import (
+    Adder,
+    Averager,
+    Clipper,
+    ConstantValue,
+    Divider,
+    FormulaStep,
+    MetricFetcher,
+    Multiplier,
+    OpenParen,
+    Subtractor,
+)
+
+class OperatorPrecedence(enum.Enum):
+    """Represents the precedence of an operator."""
+
+    Subtraction = 1
+    Addition = 2
+    Division = 3
+    Multiplication = 4
+    Primary = 9
+
+    def __gt__(self, other: OperatorPrecedence) -> bool:
+        """Return whether the precedence of this operator is greater than the precedence of the other operator.
+
+        Args:
+            other: The other operator (on the right-hand side).
+
+        Returns:
+            bool: Whether the precedence of this operator is greater than the precedence of the other operator.
+        """
+        return self.value > other.value
+
+
+    def __lt__(self, other: OperatorPrecedence) -> bool:
+        return self.value < other.value
+
+    def __le__(self, other: OperatorPrecedence) -> bool:
+        return self.value <= other.value
+
+    @staticmethod
+    def map(operator: str) -> OperatorPrecedence:
+        """Return the precedence of the operator.
+
+        Args:
+            operator: The operator.
+
+        Returns:
+            int: The precedence of the operator.
+        """
+        match operator:
+            case "+":
+                return OperatorPrecedence.Addition
+            case "-":
+                return OperatorPrecedence.Subtraction
+            case "*":
+                return OperatorPrecedence.Multiplication
+            case "/":
+                return OperatorPrecedence.Division
+            case _:
+                return OperatorPrecedence.Primary
+
+
+class StackItem:
+    """Represents a stack item for the formula printer."""
+
+    def __init__(self, value: str, precedence: OperatorPrecedence):
+        """Initialize the StackItem.
+
+        Args:
+            value: The value of the stack item.
+            precedence: The precedence of the stack item.
+        """
+        self.value = value
+        self.precedence = precedence
+
+    def __str__(self) -> str:
+        """Return the string representation of the stack item.
+
+        This is used for debugging purposes.
+
+        Returns:
+            str: The string representation of the stack item.
+        """
+        return f"(\"{self.value}\", {self.precedence})"
+
+    def plvalue(self, outer_precedence: OperatorPrecedence) -> str:
```

Comment:
by reading the function documentation, it should be clear that l=left, r=right, p=parenthesis. But I agree, longer names are always quicker to understand. :)

Reply:
I thought about that, but still couldn't understand what this is about, so maybe some documentation could also help. I can imagine this might be a common term if compilers development or when working with parsers, but from an external observers POV it is not obvious. Also the implementation of both are the same, which adds to the confusion.  :D Is this a bug? You have having the same function with 2 names for some other reason?